NEXT-12786 - [NEW] Administration: Add error-handling (#216)
shopware/docs: guides/plugins/plugins/administration/add-error-handling.md

```html
<div>
    <sw-field ... :error="propertyError"></sw-field>
</div>
```
​
### The mapErrors Service
​
Like every Vuex mapping, fetching the errors from the store may be very repetitive and error-prone.
Because of this we provide you an Vuex like mapper function:
​
```javascript
mapPropertyErrors(subject, properties)
```
​
Here, the `subject` parameter is the entity name (not the entity itself) and `properties` is an array of the properties you want to map.
You can spread its result to create computed properties in your component.
The functions returned by the mapper are named like a camelCase representation of your input, suffixed with `Error`.

This is an example from the `sw-product-basic-form` component:
​
```javascript
const { mapPropertyErrors } = Shopware.Component.getComponentHelper();
    
Component.register('sw-product-basic-form', {
    computed: {
        ...mapPropertyErrors('product', [
            'name',
            'description',
            'productNumber',
            'manufacturerId',
            'active',
            'markAsTopseller'
        ])
    }
})
```

Which then are bound to the inputs like this:

```html

> <sw-field type="text" v-model="product.name" :error="productNameError"

Author: ```suggestion
<sw-field type="text" v-model="product.name" :error="productNameError">
```

> <sw-field type="text" v-model="product.name" :error="productNameError"

Author: ```suggestion
<sw-field type="text" v-model="product.name" :error="productNameError">
```

> # API error handling in administration)

Reviewer: ```suggestion
# API error handling in administration
```

​
## Overview

The Shopware 6 Administration stores API errors in the [Vuex](https://vuex.vuejs.org/).

> Where they are centrally accessible to your components, with a flat data structure wich looks like this:

Author: I meant the Vuex store

> Where they are centrally accessible to your components, with a flat data structure wich looks like this:

Reviewer 2: Ok, in this case the suggestion is valid 👍

> Where they are centrally accessible to your components, with a flat data structure wich looks like this:

Author: i modifed the sugestion a bit are u fine with this?

​
```javascript
function getApiErrorFromPath (state) => (entityName, id, path)
```
​
Where path is an `array` representing the nested property names of your entity.

Also we provide a wrapper which can also handle nested fields in object notation, which is much easier to use for scalar fields:
​
```javascript
function getApiError(state) => (entity, field)
```
​
For example an empty product name would result in an error with the path `product.name`, instead of having the array `['product', 'name']`.

In your Vue component, use computed properties to not flood your templates with store calls.
​
```javascript
computed: {
    propertyError() {
        return this.$store.getters.getApiError(myEntity, 'myFieldName');
    },
    nestedpropertyError() {
        return this.$store.getters.getApiError(myEntity, 'myFieldName.nested');
    }
}
```

Those computed properties can then be normally used in your templates:

```html
<div>
    <sw-field ... :error="propertyError"></sw-field>
</div>
```
​
### The mapErrors Service
​
Like every Vuex mapping, fetching the errors from the store may be very repetitive and error-prone.
Because of this we provide you an Vuex like mapper function:

> ​

Author: ```suggestion
```

​
```javascript
mapPropertyErrors(subject, properties)
```
​
where subject is the entity name (not the entity itself) and properties is an array of properties you want to map.
You can spread its result to create computed properties in your component.
The functions returned by the mapper are named like a camelCase representation of your input suffixed with `Error`.

This is an example from the `sw-product-basic-form` component:
​
```javascript
const { mapPropertyErrors } = Shopware.Component.getComponentHelper();
    

> Component.register('sw-product-basic-form', {    

Reviewer: something is weird here - triple `register` ?

> Component.register('sw-product-basic-form', {    
>     Component.register('sw-product-basic-form', {
> Component.register('sw-product-basic-form', {    

Author: ```suggestion
Component.register('sw-product-basic-form', {
```

    computed: {
        ...mapPropertyErrors('product', [
            'name',
            'description',
            'productNumber',
            'manufacturerId',
            'active',
            'markAsTopseller'
        ])
    }
})
```

> Wich then are bound to the inputs like this:

Reviewer 3: Which

> Wich then are bound to the inputs like this:

Author: ```suggestion
Which then are bound to the inputs like this:
```

> Which then are bound to the inputs like this:

Author: it was the missing end tag

> Which then are bound to the inputs like this:

Reviewer 2: Ah perfect - good catch 👍

```html
<sw-field type="text" v-model="product.name" :error="productNameError"
``` 
​
### Error configuration for pages
​

> When working with nested views you need a way to tell the user that an error occurred on another view, e.g tab.

Reviewer 2: ```suggestion
When working with nested views, you need a way to tell the user that an error occurred on another view, e.g in another `tab`.
```

> For this you can write a config for your `sw-page` component which looks like: 

Reviewer 2: ```suggestion
For this you can write a config for your `sw-page` component which looks like seen below: 
```

​
```
{
  "sw.product.detail.base": {
    "product": [
      "taxId",
      "price",
      "stock",
      "manufacturerId",
      "name"
    ]
  },
  "sw.product.detail.cross.selling": {
    "product_cross_selling": [
      "name",
      "type",
      "position"
    ]
  }
}
```
​
This can then directly imported and used in the `mapPageError` computed property:

```javascript
import errorConfiguration from './error.cfg.json';

const { mapPageErrors } = Shopware.Component.getComponentHelper();

Shopware.Component.register('sw-product-detail', {
    computed: {
        ...mapPageErrors(errorConfiguration),
    }
}
```

> Wich then makes it possible to indicate if one or more errors exists, in another view or a tab:

Reviewer 2: ```suggestion
This makes it possible to indicate if one or more errors exists, in another view or a tab:
```
